File: 1396.design-underground-system.py
#
# @lc app=leetcode id=1396 lang=python3
#
# [1396] Design Underground System
#
from collections import defaultdict

# @lc code=start
# S: O(n)
class UndergroundSystem:
    # T: O(1)
    def __init__(self):
        # k: passenger_id, v: (start_station, timestamp)
        self.last_check_in = {}

        # k: (start_station, end_station), v: list of durations
        self.route_duration = defaultdict(list)  

    # T: O(1)
    def checkIn(self, id: int, stationName: str, t: int) -> None:
        self.last_check_in[id] = (stationName, t)

    # T: O(1)
    def checkOut(self, id: int, stationName: str, t: int) -> None:
        (start, t_start) = self.last_check_in.get(id, (None, None))
        if start is not None and t_start is not None:
            self.route_duration[(start, stationName)].append(t - t_start)

    # T: O(n)
    def getAverageTime(self, startStation: str, endStation: str) -> float:
        durations = self.route_duration.get((startStation, endStation), None)
        if durations is None:
            return 0
        return sum(durations) / len(durations)


# Your UndergroundSystem object will be instantiated and called as such:
# obj = UndergroundSystem()
# obj.checkIn(id,stationName,t)
# obj.checkOut(id,stationName,t)
# param_3 = obj.getAverageTime(startStation,endStation)
# @lc code=end

File: 1396.design-underground-system.cpp
/*
 * @lc app=leetcode id=1396 lang=cpp
 *
 * [1396] Design Underground System
 */
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>

using namespace std;

// @lc code=start
typedef pair<string, int> check_in; // station and check-in time

// S: O(n)
class UndergroundSystem {
    unordered_map<int, check_in> passenger_check_in;  // k: passenger_id, v: check_in (station, time)
    unordered_map<string, vector<int>> route_durations;  // k: station1_station2, v: list of durations

public:
    UndergroundSystem() {}
    
    // O(1)
    void checkIn(int id, string stationName, int t) {
        passenger_check_in[id] = {stationName, t};
    }
    
    // O(1)
    void checkOut(int id, string stationName, int t) {
        const auto& [pre_station, t_pre] = passenger_check_in[id];
        route_durations[pre_station + "=>" + stationName].push_back(t - t_pre);
    }
    
    // O(n)
    double getAverageTime(string startStation, string endStation) {
        double res = 0.;
        const string route = startStation + "=>" + endStation;
        for (const auto& t : route_durations[route]) {
            res += t;
        }
        return res / route_durations[route].size();
    }
};

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem* obj = new UndergroundSystem();
 * obj->checkIn(id,stationName,t);
 * obj->checkOut(id,stationName,t);
 * double param_3 = obj->getAverageTime(startStation,endStation);
 */
// @lc code=end

